core/services/gateway/handlers/webapicapabilities/handler.go

```diff
@@ -1,6 +1,116 @@
 package webapicapabilities
 
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.uber.org/multierr"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/gateway/api"
+	"github.com/smartcontractkit/chainlink/v2/core/services/gateway/config"
+	"github.com/smartcontractkit/chainlink/v2/core/services/gateway/handlers"
+)
+
 const (
 	// NOTE: more methods will go here. HTTP trigger/action/target; etc.
-	MethodWebAPITarget = "web_api_target"
+	MethodWebAPITarget  = "web_api_target"
+	MethodWebAPITrigger = "web_api_trigger"
 )
+
+type handler struct {
+	config         HandlerConfig
+	donConfig      *config.DONConfig
+	don            handlers.DON
+	savedCallbacks map[string]*savedCallback
+	mu             sync.Mutex
+	lggr           logger.Logger
+}
+
+type HandlerConfig struct {
+	MaxAllowedMessageAgeSec uint
+}
+type savedCallback struct {
+	id         string
+	callbackCh chan<- handlers.UserCallbackPayload
+}
+
+var _ handlers.Handler = (*handler)(nil)
+
+func NewWorkflowHandler(handlerConfig json.RawMessage, donConfig *config.DONConfig, don handlers.DON, lggr logger.Logger) (*handler, error) {
+	var cfg HandlerConfig
+	err := json.Unmarshal(handlerConfig, &cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &handler{
+		config:         cfg,
+		donConfig:      donConfig,
+		don:            don,
+		savedCallbacks: make(map[string]*savedCallback),
+		lggr:           lggr.Named("WorkflowHandler." + donConfig.DonId),
+	}, nil
+}
+
+func (d *handler) HandleUserMessage(ctx context.Context, msg *api.Message, callbackCh chan<- handlers.UserCallbackPayload) error {
+	d.mu.Lock()
+	d.savedCallbacks[msg.Body.MessageId] = &savedCallback{msg.Body.MessageId, callbackCh}
+	don := d.don
+	d.mu.Unlock()
+	body := msg.Body
+	var payload TriggerRequestPayload
+	err := json.Unmarshal(body.Payload, &payload)
+	if err != nil {
+		d.lggr.Errorw("error decoding payload", "err", err)
+		callbackCh <- handlers.UserCallbackPayload{Msg: msg, ErrCode: api.UserMessageParseError, ErrMsg: fmt.Sprintf("error decoding payload %s", err.Error())}
```

Comment:
do we need to write to callbackCh here? returning an `err` from this function seems sufficient to send a HTTP Error back to the client. `callbackCh` seems to be only needed to handle messages from the node in `HandleNodeMessage`
similar questions for other error scenarios in this function.

Reply:
I want the http connection to have the full error message, as that's the only way the gateway can respond.   Good catch though, I want to return no error here and the UserResponse so that the ErrMsg is used.
 The gateway code does
```
// send to the handler
	responseCh := make(chan handlers.UserCallbackPayload, 1)
	err = handler.HandleUserMessage(ctx, msg, responseCh)
	if err != nil {
		return newError(g.codec, msg.Body.MessageId, api.HandlerError, err.Error())
	}
	// await response
	var response handlers.UserCallbackPayload
	select {
	case <-ctx.Done():
		return newError(g.codec, msg.Body.MessageId, api.RequestTimeoutError, "handler timeout")
	case response = <-responseCh:
		break
	}
	if response.ErrCode != api.NoError {
		return newError(g.codec, msg.Body.MessageId, response.ErrCode, response.ErrMsg)
	}
	// encode
	rawResponse, err = g.codec.EncodeResponse(response.Msg)
	if err != nil {
		return newError(g.codec, msg.Body.MessageId, api.NodeReponseEncodingError, "")
	}
	promRequest.WithLabelValues(api.NoError.String()).Inc()
	return rawResponse, api.ToHttpErrorCode(api.NoError)
```